src/restruct_poc/math/vector_operations.hpp

```diff
@@ -18,12 +20,54 @@ KOKKOS_INLINE_FUNCTION void VecTilde(VectorType vector, MatrixType matrix) {
     matrix(2, 2) = 0.;
 }
 
-constexpr std::array<double, 3> CrossProduct(std::array<double, 3> a, std::array<double, 3> b) {
-    auto c = std::array<double, 3>{};
-    c[0] = a[1] * b[2] - a[2] * b[1];
-    c[1] = a[2] * b[0] - a[0] * b[2];
-    c[2] = a[0] * b[1] - a[1] * b[0];
-    return c;
+/// Calculate the dot product between two vector views
+template <typename VectorType>
+KOKKOS_INLINE_FUNCTION double DotProduct(VectorType a, VectorType b) {
+    double sum = 0.;
+    for (int i = 0; i < a.extent_int(0); ++i) {
+        sum += a(i) * b(i);
+    }
+    return sum;
+}
+
+/// Calculate the dot product between two vector views
+constexpr double DotProduct(Array_3 a, Array_3 b) {
+    double sum = 0.;
+    for (int i = 0; i < 3; ++i) {
+        sum += a[i] * b[i];
+    }
+    return sum;
+}
+
+/// Calculate the cross product between two vector views
+template <typename VectorType>
+KOKKOS_INLINE_FUNCTION void CrossProduct(VectorType a, VectorType b, VectorType c) {
+    c(0) = a(1) * b(2) - a(2) * b(1);
+    c(1) = a(2) * b(0) - a(0) * b(2);
+    c(2) = a(0) * b(1) - a(1) * b(0);
+}
+
+/// Calculate the cross product between two vectors
+constexpr Array_3 CrossProduct(const Array_3& a, const Array_3& b) {
+    return Array_3{
+        a[1] * b[2] - a[2] * b[1],
+        a[2] * b[0] - a[0] * b[2],
+        a[0] * b[1] - a[1] * b[0],
+    };
+}
+
+/// UnitVector returns the unit vector of the given vector
+constexpr Array_3 UnitVector(const Array_3& v) {
+    // Calculate vector norm
+    double norm = std::sqrt(v[0] * v[0] + v[1] * v[1] + v[2] * v[2]);
+    if (norm == 0) {
+        throw std::invalid_argument("Cannot normalize a zero vector");
+    }
```

Comment:
Would it not be correct to just return v in this case? 

Also, I'm surprised we're not throwing a warning/error comparing a double to an integer here.

Reply:
It wouldn't be a proper unit vector without having a magnitude of 1. I think the downstream logic would fail if the vector were zero. I'm also surprised that there's no issue with the integer comparison, probably some sort of automatic type promotion. However, I can easily fix it.